external/storm-hdfs-oci/src/main/java/org/apache/storm/container/oci/LocalOrHdfsImageTagToManifestPlugin.java

```diff
@@ -0,0 +1,294 @@
+        if (!fs.exists(imageToHash)) {
+            String message = "Could not load hdfs image to hash file, " + hdfsImageToHashFile + " doesn't exist";
+            LOG.error(message);
+            throw new IOException(message);
+        }
+
+        long newHdfsModTime = fs.getFileStatus(imageToHash).getModificationTime();
+        if (newHdfsModTime == hdfsModTime) {
+            LOG.debug("Did not load hdfs image to hash file, file is unmodified");
+            return null;
+        }
+        hdfsModTime = newHdfsModTime;
+
+        return new BufferedReader(new InputStreamReader(fs.open(imageToHash)));
+    }
+
+    // You may specify multiple tags per hash all on the same line.
+    // Comments are allowed using #. Anything after this character will not
+    // be read
+    // Example file:
+    // foo/bar:current,fizz/gig:latest:123456789
+    // #this/line:wont,be:parsed:2378590895
+    //
+    // This will map both foo/bar:current and fizz/gig:latest to 123456789
+    private static Map<String, String> readImageToHashFile(BufferedReader br) throws IOException {
+        if (br == null) {
+            return null;
+        }
+
+        String line;
+        Map<String, String> imageToHashCache = new HashMap<>();
+        while ((line = br.readLine()) != null) {
+            int index;
+            index = line.indexOf("#");
+            if (index == 0) {
+                continue;
+            } else if (index != -1) {
+                line = line.substring(0, index);
+            }
+
+            index = line.lastIndexOf(":");
+            if (index == -1) {
+                LOG.warn("Malformed imageTagToManifest entry: " + line);
+                continue;
+            }
+            String imageTags = line.substring(0, index);
+            String[] imageTagArray = imageTags.split(",");
+            String hash = line.substring(index + 1);
+
+            if (!hash.matches(ALPHA_NUMERIC) || hash.length() != SHA256_HASH_LENGTH) {
+                LOG.warn("Malformed image hash: " + hash);
+                continue;
+            }
+
+            for (String imageTag : imageTagArray) {
+                imageToHashCache.put(imageTag, hash);
+            }
+        }
+        return imageToHashCache;
+    }
+
+
+    @Override
+    public synchronized ImageManifest getManifestFromImageTag(String imageTag) throws IOException {
+        String hash = getHashFromImageTag(imageTag);
+        ImageManifest manifest = manifestCache.get(hash);
+        if (manifest != null) {
+            return manifest;
+        }
+        Path manifestPath = new Path(manifestDir + hash);
+        FileSystem fs = manifestPath.getFileSystem(new Configuration());
+        FSDataInputStream input;
+        try {
+            input = fs.open(manifestPath);
+        } catch (IllegalArgumentException iae) {
+            throw new IOException("Manifest file is not a valid HDFS file: "
+                + manifestPath.toString(), iae);
```

Comment:
The manifestPath seems to be a local file not HDFS.

Reply:
This is on HDFS.
see
```
manifestDir = ObjectReader.getString(conf.get(DaemonConfig.STORM_OCI_IMAGE_HDFS_TOPLEVEL_DIR)) + "/manifests/";
```